scipy/stats/_discrepancy.pyx

```diff
@@ -0,0 +1,382 @@
+    cdef:
+        Py_ssize_t n = sample_view.shape[0]
+        Py_ssize_t d = sample_view.shape[1]
+        Py_ssize_t i = 0, j = 0, k = 0
+        double x_kikj, prod = 1, disc = 0
+
+    for i in range(n):
+        for j in range(n):
+            for k in range(d):
+                x_kikj = fabs(sample_view[i, k] - sample_view[j, k])
+                prod *= 3.0 / 2.0 - x_kikj + x_kikj ** 2
+            disc += prod
+            prod = 1
+
+    if iterative:
+        n += 1
+
+    return - (4.0 / 3.0) ** d + 1.0 / (n ** 2) * disc
+
+
+cdef double mixture_discrepancy(double[:, ::1] sample_view,
+                             bint iterative):
+    cdef:
+        Py_ssize_t n = sample_view.shape[0]
+        Py_ssize_t d = sample_view.shape[1]
+        Py_ssize_t i = 0, j = 0, k = 0
+        double prod = 1, disc = 0, disc1 = 0
+
+    for i in range(n):
+        for j in range(d):
+                prod *= (
+                        5.0 / 3.0 - 0.25 * fabs(sample_view[i, j] - 0.5)
+                        - 0.25 * fabs(sample_view[i, j] - 0.5) ** 2
+                )
+        disc1 += prod
+        prod = 1
+
+    cdef:
+        double disc2 = 0
+
+    for i in range(n):
+        for j in range(n):
+            for k in range(d):
+                    prod *= (15.0 / 8.0
+                             - 0.25 * fabs(sample_view[i, k] - 0.5)
+                             - 0.25 * fabs(sample_view[j, k] - 0.5)
+                             - 3.0 / 4.0 * fabs(sample_view[i, k]
+                                                - sample_view[j, k])
+                             + 0.5
+                             *fabs(sample_view[i, k] - sample_view[j, k]) ** 2)
+            disc2 += prod
+            prod = 1
+
+    if iterative:
+        n += 1
+
+    disc = (19.0 / 12.0) ** d
+    disc1 = 2.0 / n * disc1
+    disc2 = 1.0 / (n ** 2) * disc2
+
+    return disc - disc1 + disc2
+
+
+cdef double l2_star_discrepancy(double[:, ::1] sample_view,
+                             bint iterative):
+    cdef:
+        Py_ssize_t n = sample_view.shape[0]
+        Py_ssize_t d = sample_view.shape[1]
+        Py_ssize_t i = 0, j = 0, k = 0
+        double prod = 1, disc1 = 0
+
+    for i in range(n):
+        for j in range(d):
+                prod *= 1 - sample_view[i, j] ** 2
+
+        disc1 += prod
+        prod = 1
+
+    cdef:
+        double disc2 = 0, tmp_sum = 0
+
+    for i in range(n):
+        for j in range(n):
+            for k in range(d):
+                    prod *= (
+                            1 - max(sample_view[i, k], sample_view[j, k])
+                    )
+            tmp_sum += prod
+            prod = 1
+
+        disc2 += tmp_sum
+        tmp_sum = 0
+
+    if iterative:
+        n += 1
+
+    cdef double one_div_n = <double> 1 / n
+    return sqrt(
+        pow(3, -d) - one_div_n * pow(2, 1 - d) * disc1 + 1 / pow(n, 2) * disc2
+    )
+
+
+def update_discrepancy(x_new, sample, double initial_disc):
```

Comment:
Same here, I would put this back in `_qmc`.

Reply:
Since I am calling `cdef` inside `update_discrepancy` and `discrepancy` function, I cannot put this function in `_qmc`. Take a look to this [link](https://notes-on-cython.readthedocs.io/en/latest/classes.html). Apparently, `cdef` is much faster than `coded`. That's why I am letting this here

Reply:
Sure `cdef` is faster than `cpdef` as with `cpdef` you basically have 2 versions, a python wrapper and the c one. So there is an overhead to do handle the logic. But as the function is complex, what really matters in the end is not really the way we call it as most of the CPU is going to be used inside the function itself.

You can leave this like that for now, and we can see in the end what other think.